### PR TITLE
refactor fragment shader

### PR DESCRIPTION
**Pull request by RobertGemmaJr**
Some more general refactoring of the shader code. Includes some changes made in preparation for blending models in the shader but we're no longer going that direction. 

I undid the modelData struct so code is now refactored in  a way that's logical for the THREE.js blending we're now going to do. I figure it's best to get these changes in now while it's relatively small, working on the next steps in another branch

### CODE REVIEW

**Review by eldu — COMMENTED** [review]
submitting this review before a meeting... 

I'll reread create_model again after my meeting

File: src/utils/index.js [hunk 1]
```diff
@@ -35,7 +35,6 @@ function getAframeModels(models) {
 }
 
 // Recursively find the differences between two objects
-// https://davidwells.io/snippets/get-difference-between-two-objects-javascript
 function deepDifference(oldObj, newObj) {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/utils/index.js]: The function is an expansion loadash's `isEqual`. It uses `isEqual`, `isArray`, `isObject`, and `transform` from loadash. It recursively checks properties that are arrays or objects, I believe `isEqual` only checks the first level.

Either way I added the link comment back, I'm not sure why I ever deleted that

File: src/Aframe/volume.js [hunk 2]
```diff
@@ -343,17 +329,24 @@ AFRAME.registerComponent("volume", {
...
-        ? // TEMP - use first material (GH issue #68)
-          new RawShaderMaterial({
-            ...DEFAULT_MATERIAL,
-            uniforms: this.uniforms.values().next().value,
-          })
-        : // No models - use default material
-          new RawShaderMaterial(DEFAULT_MATERIAL);
+  // Pass array of models' data into the shader
+  updateModels: function () {
+    console.log("MODELS LOADED", this.modelsData);
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/Aframe/volume.js:334]: I'll delete it, easy enough to re-add if I need it down the line

File: src/utils/index.js [hunk 3]
```diff
@@ -53,7 +52,7 @@ function deepDifference(oldObj, newObj) {
   return changes(newObj, oldObj);
 }
 
-function validateVec3String(props, propName, componentName) {
+const validateVec3String = function (props, propName, componentName) {
```
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/utils/index.js]: It definitely shouldn't be const and a function. I changed it to `function validateVec3String(props, propName, componentName) {` to match the other validation functions for now

Is it more typical to use fat arrow notation even if a separate file? I'm happy to change

**Review by kmilo9999 — APPROVED** [review]
I reviewed the wrong PR.  This one is fine with the suggested changes. For the record : "the sample1D is because "NVIDIA GLSL". WebGL is based on OpenGL ES 2.0. "

File: src/utils/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (eldu) [reply src/utils/index.js]: Well the arrow notation is just more modern js introduced in ES6. 

I'd likely write something like
```js
const validateVec3String = (props, propName, componentName) => { 
   ... 
}
```

Here are some blog posts

https://www.w3schools.com/js/js_arrow_function.asp
https://www.programiz.com/javascript/arrow-function

File: src/Aframe/fragment-shader.frag [hunk 4]
```diff
@@ -111,43 +142,7 @@ void main() {
...
 
-    // Loop through the volume
-    for(float i = t_start; i < t_end; i += step_size) {    
-        vec4 volumeSample = sampleAs3DTexture(u_data, data_position);
-        if (blending == 1) volumeSample = volumeSample.rrrr;
-        else if (blending == 2) volumeSample = volumeSample.gggg;
-        else if (blending == 3) volumeSample = volumeSample.bbbb;
-        else if (blending == 4) volumeSample = volumeSample.aaaa;
-        else if (blending == 5) volumeSample = volumeSample;
-        else {
```
Comment (eldu) [inline src/Aframe/fragment-shader.frag:127]: So in the old code looks like blending was just choosing the rgb channel we use. What is it used for in the refactor?

Also, in the new code, is it just what's in this `else`? So we are using the `volumeSample.rgb` as is and then doing calculations to figure out `volumeSample.a`

File: src/Aframe/fragment-shader.frag [hunk 5]
```diff
@@ -60,6 +60,37 @@ vec2 intersectBox(vec3 camera, vec3 direction, vec3 clip_min, vec3 clip_max ) {
...
+// Starting from the entry point, march the ray through the volume and sample it
+vec4 create_model(float t_start, float t_end, vec3 data_position, vec3 ray_direction) {
+    vec4 vFragColor = vec4(0, 0, 0, 0);
+
+    // Loop from t_start to t_end by step_size
+    for(float t = t_start; t < t_end; t += step_size) {
+        vec4 volumeSample = sampleAs3DTexture(model_texture, data_position);
+
+        // Initialize alpha as the max between the 3 channels
+        // volumeSample .r .g and .b are all the same exact values. Don't know what .a is supposed to be
```
Comment (eldu) [inline src/Aframe/fragment-shader.frag:72]: Looks like later on it's used to calculate the opacity. But i'm also a bit confused...

**Review by eldu — APPROVED** [review]
Alright I added in more comments, some questions more so to help me understand what's happening.

looks good to me overall

File: src/utils/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (eldu) [reply src/utils/index.js]: Oh i see, this is returning what the differences between the objects are. 

lodash's `isEqual` is a deep comparison.
https://www.samanthaming.com/tidbits/33-how-to-compare-2-objects/

File: src/utils/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (RobertGemmaJr) [reply src/utils/index.js]: For sure, I refactor all the utility functions to use fat arrow notation

File: src/Aframe/fragment-shader.frag (hunk 4 above)
Comment (RobertGemmaJr) [reply src/Aframe/fragment-shader.frag:127]: Yes, it use to be `channel` (not `blending`) so the user could select the r, g, and b channels individually if they wanted. What we're moving towards is the user being able to select with blending type they want (additive, subtractive, etc). That functionality hasn't been added yet.

**Pull request merged**